pkg/compiler: add support for len target argument for array type

dokyungs opened the pull request:
This CL introduces a special argument type "len" that describes the size
of an array in terms of the size of another array. The 2nd argument of
array type becomes a custom type, which supports either an integer range
or the new "len" argument type.

dvyukov commented on pkg/compiler/types.go:
```diff
@@ -166,19 +174,21 @@ var typeArray = &typeDesc{
-				Kind:       bufKind,
-				RangeBegin: begin,
-				RangeEnd:   end,
+				TypeCommon:   base.TypeCommon,
```
I think we should restrict this functionality only to ArrayType for now. I will make things in prog package simpler. When/if it will be needed for BufferType and we have it working for ArrayType, we can extend this to BufferType too. But I would avoid this unnecessary complexity for now.

dokyungs replied:
Reverted BufferType changes

dvyukov commented on pkg/compiler/types.go:
```diff
@@ -793,6 +803,29 @@ var typeArgSizeRange = &typeArg{
+	},
+	Check: func(comp *compiler, t *ast.Type) {
+		if t.Ident == "len" {
+			if len(t.Args) != 1 {
```
We also need to check that when it's not "len" number of args is equal to 0.

On the second though, it's actually an oversight that we allow subargs for args at all.
I've fixed one case that violated this:
https://github.com/google/syzkaller/commit/1be453f07bb52417650d9d617d41a834ecf1a0ab
and added the check:
https://github.com/google/syzkaller/commit/7a88b14122b5b70ed04bcd75932f411256e1662b

Please rebase and set MaxArgs=1 for this arg. But you still need to check that it's 0 for non-len case.

dokyungs replied:
Both done.

dvyukov commented on pkg/compiler/types.go:
```diff
@@ -793,6 +803,29 @@ var typeArgSizeRange = &typeArg{
+		}
+	},
+	Check: func(comp *compiler, t *ast.Type) {
+		if t.Ident == "len" {
```
Also check that t.HasColon is false for len.
Also check that subarg is identifier, otherwise it is possible to write `array[int8, len["foo"]]`.

dokyungs replied:
Done

dokyungs commented on prog/types.go:
```diff
@@ -350,13 +350,14 @@ const (
+	Kind         BufferKind
+	RangeBegin   uint64 // for BufferBlobRange kind
+	RangeEnd     uint64 // for BufferBlobRange kind
+	RangeLiteral string
```
Changed it to `LenTarget`

dokyungs commented on pkg/compiler/types.go:
```diff
@@ -152,8 +155,13 @@ var typeArray = &typeDesc{
 		if len(args) > 1 {
-			kind, begin, end = prog.ArrayRangeLen, args[1].Value, args[1].Value2
+			if args[1].Ident == "len" {
+				literal = args[1].Args[0].Ident
```
Added `ArrayLenTarget`

dokyungs replied:
BTW, how do we go about generating arrays of the `ArrayLenTarget` kind? I saw that `LenType` arguments are fixed up in `assignSizes` as another pass separate from argument generation.

Should we somehow make sure that generateArgs populates the target array first, so that the sizes of those arrays can be found by `ArrayLenTarget` arrays?

dvyukov replied:
I don't have precise answer.
I think we should restrict len to refer to a _preceding_ array only. Then memorize array sizes in a map as we generate fields, and then when we generate a parallel array consult that map. Something along these lines.

dokyungs replied:
Ok, I got the idea. I will ping you once I have a working version.

dokyungs commented on pkg/compiler/types.go:
```diff
@@ -152,8 +155,13 @@ var typeArray = &typeDesc{
 	Gen: func(comp *compiler, t *ast.Type, args []*ast.Type, base prog.IntTypeCommon) prog.Type {
 		elemType := comp.genType(args[0], "", base.ArgDir, false)
 		kind, begin, end := prog.ArrayRandLen, uint64(0), uint64(0)
+		literal := ""
```
Done

dokyungs commented on pkg/compiler/consts.go:
```diff
@@ -68,7 +68,8 @@ func (comp *compiler) extractConsts() map[string]*ConstInfo {
 				for i, arg := range args {
-					if desc.Args[i].Type.Kind == kindInt {
+					if desc.Args[i].Type.Kind == kindInt ||
+						(desc.Args[i].Type == typeArgSizeRangeLiteral && arg.Ident != "len") {
```
Done

dokyungs commented:
I added `arraySizes` in `state` struct, which is used to remember the sizes of all generated arrays. This map is looked up when generating `ArrayLenTarget` type arrays.

To always ensure that the len target precedes the array that refers to it, I added a compile time check in `checkArraySizeLenTarget`. Can you please take a look?

dvyukov commented on prog/rand.go:
```diff
@@ -690,7 +690,12 @@ func (a *ArrayType) generate(r *randGen, s *state) (arg Arg, calls []*Call) {
+		count = s.arraySizes[a.LenTarget]
 	}
+
+	s.arraySizes[a.FldName] = count
```
I see that checkArraySizeLenTarget requires target to be in the same struct, right? So perhaps arraySizes should be even per-struct. That would also fix the example above.
randGen.generateArgs generates a single struct, so I think we should re-create the map there.

dokyungs commented:
I will reopen this PR when I resume working on this.
